source/module_basis/module_ao/parallel_2d.h

```diff
@@ -12,117 +13,112 @@
-
-    /// create the 'comm_2D' stratege.
-    void mpi_create_cart(const MPI_Comm& diag_world);
-
-    /// set the map from local index to global index,
-    /// and set local sizes (nrow, ncol, nloc) by the way
-    int set_local2global(const int& M_A/**< global row size*/,
-        const int& N_A/**< global col size*/,
-        std::ofstream& ofs_running,
-        std::ofstream& ofs_warning);
-
-    ///@brief set the desc[9] of the 2D-block-cyclic distribution
-    void set_desc(const int& gr/**< global row size*/,
-        const int& gc/**< global col size*/,
-        const int& lld/**< leading local dimension*/,
-        bool first_time = true/**< true: call `Cblacs_get`; false: use `this->blacs_ctxt`*/);
-#else
-    void set_serial(const int& M_A/**< global row size*/,
-        const int& N_A/**< global col size*/);
+    /**
+     * @brief Initialize a BLACS grid with the given MPI communicator
+     * and set up the info of a block-cyclic distribution.
+     *
+     */
+    int init(
+        const int mg,
+        const int ng,
+        const int nb, // square block is assumed
+        const MPI_Comm comm,
+        bool mode = 0
+    );
+
+    /**
+     * @brief Set up the info of a block-cyclic distribution using given
+     * MPI communicator and BLACS context.
+     *
+     */
+    int set(
+        const int mg,
+        const int ng,
+        const int nb, // square block is assumed
+        const MPI_Comm comm_2D,
+        const int blacs_ctxt
+    );
+
+    /// BLACS context
+    int blacs_ctxt = -1;
+
+    /// ScaLAPACK descriptor
+    int desc[9] = {};
+
+    /// 2D Cartesian MPI communicator
+    MPI_Comm comm_2D = MPI_COMM_NULL;
 #endif
 
-    void set_global2local(const int& M_A,
-        const int& N_A,
-        const bool& div_2d,
-        std::ofstream& ofs_running);
+    void set_serial(const int mg, const int ng);
+
+    // FIXME the following variables should be private, but they are
+    // widely used in the code currently. Public visibility is kept
+    // for now, but might be changed in the future.
+
+    /// local size (nloc = nrow * ncol)
+    int nrow = 0;
+    int ncol = 0;
+    int64_t nloc = 0;
+
+    /// block size
+    int nb = 1;
+
+    /// number of processes in each dimension of the MPI Cartesian grid
+    int dim0 = 0;
+    int dim1 = 0;
+
+    /// process coordinate in the MPI Cartesian grid
+    int coord[2] = {-1, -1};
+
+    /// test parameter
+    int testpb = 0;
+
 
 protected:
 
     /// map from global index to local index
-    int* global2local_row_ = nullptr;
-    int* global2local_col_ = nullptr;
+    std::vector<int> global2local_row_;
+    std::vector<int> global2local_col_;
 
     /// map from local index to global index
-    std::vector<int> local2global_row_;				// Peize Lin change int* to vector 2022.08.03
+    std::vector<int> local2global_row_;
     std::vector<int> local2global_col_;
+    // Peize Lin change int* to vector 2022.08.03
```

Comment:
I think this line can be deleted.

Reply:
This line has been removed in the latest commit.